### PR TITLE
[finance_report_infra] feat(runtime): close all remaining enforcement gaps — probes + smoke parity + real-storage pipeline + input-keyed cassette

### LINKED ISSUE #1520: Critical test gap: upload→S3→load→LLM-parse pipeline never runs the real StorageService (use boto3 + in-memory S3, not MinIO)
## Problem
The core user journey — **upload a statement → store to S3 → load it back → parse it** — is **never exercised end-to-end by any counted test**. Every seam is fully stubbed, so the suite stays green even if the real `StorageService`/boto3 code path is broken.

## Evidence (current behaviour)
- **Upload tests** (`tests/api/test_statements_router.py`): `storage_stub` swaps `StorageService` for a hand-written `DummyStorage` (upload→`None`, get→`b"dummy content"`). The **real boto3 code never runs**.
- **Parse tests** (`tests/extraction/test_statement_parsing_audit_logging.py`): `monkeypatch` replaces **both** `StorageService` (file load) **and** `ExtractionService.parse_document` (LLM) with fakes.
- **`tests/extraction/test_storage.py`**: `patch("boto3.client")` with a `MagicMock` — asserts \*that\* `put_object` was called, but never actually stores/retrieves bytes (no functional round-trip).
- **e2e bypasses it**: `tests/e2e/conftest.py::seeded_parsed_statement` *"bypasses the LLM/OCR provider … Fixture-seeded, already-parsed statement"* — seeds a pre-parsed row straight into the DB.
- The synthetic PDF fixtures (`tests/fixtures/generated/*.pdf` + `*_expected.json`, built "for E2E testing") are only fed to the extraction layer in isolation — never through a storage round-trip.

## Why critical
Green CI does **not** prove a user can upload a statement and get it parsed. A regression in the real `StorageService`/boto3 path, the upload→store→load→parse wiring, or the `aioboto3` vs `boto3` split ships undetected.

## Approach — boto3 + in-memory S3 simulation (NOT real MinIO)
The app is **S3-backend-agnostic** (boto3 against any S3-compatible backend). A real backend in CI adds **zero** value over an in-memory simulation and is slower. So:

- **Do NOT** add a MinIO/LocalStack service container.
- **DO** use an in-memory S3 simulation through boto3 — e.g. `moto` (`mock_aws`) — so the **real `StorageService` and real boto3 calls execute** against a fast in-memory store. No network, no service, runs in the normal (fast) test path.

This replaces today's full `DummyStorage` stub with a *functional* in-memory S3, turning the upload/parse tests into a genuine pipeline test while staying fast.

## Proposed test
New test (can live in the normal unit/integration path — no service container):
1. Activate the in-memory S3 sim (e.g. `moto`'s `mock_aws`); create the bucket.
2. `POST /statements/upload` with a committed fixture (e.g. `dbs_statement_fixture.pdf`) → **real** `StorageService.upload_bytes` → in-memory S3.
3. Pipeline loads the object back → **real** `StorageService.get_object` → in-memory S3 (real round-trip: bytes written == bytes read).
4. `ExtractionService` runs with an **LLM cassette** (recorded; CI never calls a real LLM — correct).
5. Assert the parsed result equals `dbs_statement_fixture_expected.json`.

## Acceptance criteria
- [ ] AC-1: a test uploads a fixture PDF through `/statements/upload` and reads it back via the **real** `StorageService` against an **in-memory boto3 S3** (no `DummyStorage`, no `StorageService` monkeypatch, no real MinIO).
- [ ] AC-2: the bytes stored equal the bytes read back (functional round-trip, not a call-assertion mock).
- [ ] AC-3: the same test parses via an LLM cassette and asserts output == the committed `*_expected.json`.
- [ ] AC-4: the test stays in the fast path (no service container; comparable runtime to existing unit tests).
- [ ] AC-5: breaking the real storage round-trip (e.g. dropping the `upload_bytes`/`get_object` wiring) makes this test **fail** — proving it intercepts.
- [ ] AC-6: real-pipeline coverage (real `StorageService` + parse orchestration) is reflected in the coverage gate.

## Notes
- Keep the existing fast mocked unit tests; this adds a functional in-memory pipeline test, it doesn't replace them.
- Explicitly **not** in scope: a real MinIO/LocalStack service in CI (the backend is interchangeable; boto3 + in-memory sim is the fast, sufficient choice).
- `moto` is the likely lib (none is in deps today); confirm async support if the path touches `aioboto3` (`StorageService` is sync boto3; `Bootloader._check_s3` is aioboto3 — out of scope here).

### LINKED ISSUE #1578: [finance_report_infra] runtime: smoke ↔ declaration parity (invariant 6) + tag→staging-smoke gate
## Context

`common/runtime` invariant 6: **the smoke test asserts exactly the declared set for its tier** (assertion count == declared count), and a release tag promotes only after the staging smoke (real backends, real LLM) passes. Neither is enforced today — `tools/smoke_test.sh` and the manifest can drift silently.

## Acceptance

- A parity check binds the smoke test's asserted dependency set to `DEPENDENCY_MANIFEST.required_for(tier)` — adding/removing a manifest entry without updating the smoke coverage fails CI.
- The release-tag promotion path is gated on the staging smoke passing.

Tracked in `common/runtime/todo.md`. Depends on #1577 (manifest-driven validate). Follows the retired umbrella #1554.

### LINKED ISSUE #1580: [finance_report_infra] runtime: presence probes for the 5 declared-but-unprobed dependencies (cache, workflow_engine, telemetry, analytics, market_data)
## Context

`DEPENDENCY_MANIFEST` declares 8 external dependencies, but only 3 have `DependencyCheck` adapters today (`DatabaseCheck`, `ObjectStorageCheck`, `LlmCheck`). The remaining 5 — Redis (`cache`), Prefect (`workflow_engine`), OTel (`telemetry`), OpenPanel (`analytics`), Yahoo (`market_data`) — are declaration-only: no probe, no presence assertion in their required tiers.

## Acceptance

- Each of the 5 gets a `DependencyCheck` adapter (in `apps/backend/src/runtime/extension/adapters.py`, or in its owning domain package implementing the port — dependency inversion per the charter).
- Probes return binary `PRESENT`/`ABSENT` `ProbeResult`s (no `skipped`) and are wired into the tier-appropriate gate once #1577 lands.

Tracked in `common/runtime/todo.md`. Follows the retired umbrella #1554.

### LINKED ISSUE #1581: [finance_report_infra] runtime: LLM recording is input-keyed (invariant 5) — changed input ⇒ recording miss
## Context

`common/runtime` invariant 5 for model-dominant dependencies: CI replays an **input-keyed** recording — a changed input must produce a recording miss (fail), never a stale-but-passing replay. The cassette gate (#1563) pinned the real GLM response shape in plain CI, but the recording is not yet keyed by input.

## Acceptance

- The LLM cassette lookup is keyed on the request input (prompt/model/params hash); a changed input misses the recording and fails CI instead of replaying a stale response.
- The staging gate continues to exercise the real provider and does not read recordings.

Tracked in `common/runtime/todo.md` ("Substitutes", invariants 4/5 — the S3 half is #1520). Follows the retired umbrella #1554.

### PR DESCRIPTION
**Pull request by wangzitian0**
Closes #1580. Closes #1578. Closes #1520. Closes #1581.

One PR closing every remaining runtime-enforcement issue, as planned. TDD throughout (each piece red→green).

## #1580 — probes for the five declared-but-unprobed dependencies (AC-runtime.4.1)

- `RedisCheck` (raw TCP `PING`, zero new deps), `WorkflowEngineCheck` (Prefect `/health`), `TelemetryCheck` (OTLP TCP connect, gRPC/HTTP-agnostic), `AnalyticsCheck` (OpenPanel HTTP, <500 = present), `MarketDataCheck` (Yahoo chart endpoint; 429 = present — presence ≠ quota). All binary PRESENT/ABSENT, never raise, "not configured" = ABSENT.
- Wired into `Bootloader._required_checks` → **invariant 2 (absent ⇒ fail) is now enforceable for the whole manifest**; the #1577 warning path is empty and guarded by `test_every_tier_declaration_is_smoke_assertable`.
- Two **over-declarations corrected**: `cache` and the REAL `llm` provider are staging/prod requirements — the preview stack has no Redis (only `10.app` renders `REDIS_URL`), and CI/preview run the input-keyed cassette substitute (invariant 5), which needs no key.

## #1578 — smoke ↔ declaration parity, invariant 6 (AC-runtime.6.1)

- `GET /health?full=1` probes `DEPENDENCY_MANIFEST.required_for(resolve_env_tier(...))` and returns 503 on any absence; response carries `tier`. The plain `/health` stays the light Docker-healthcheck form (db+s3) — and its stale `"skipped"` acceptance is gone.
- `tools/smoke_test.sh` now asserts `Dependency Presence (manifest, full)` via that URL; a test pins the line so it can't be dropped. Works across preview/staging/prod because the assertion follows the runtime's own tier.
- The **tag→production gate already requires the staging smoke** (`release.yml` verifies a successful staging deploy on the tag) — verified, documented in todo.md rather than rebuilt.

## #1520 — real upload→S3→load→parse pipeline (AC-runtime.8.1)

- `tests/api/test_statement_pipeline_moto.py`: the **real `StorageService`/boto3** (incl. real `_ensure_bucket`, signing, round-trip) against **moto in-memory S3** through `POST /statements/upload` — byte-identical store/read-back; the loaded-back statement parses via the real `stream_ai_json` transport **replaying a committed REAL-GLM cassette**, asserted equal to the committed `text_statement_expected.json`. Fast path, no service container. `moto[s3]` added to dev deps.
- The `DummyStorage` unit tests stay (cheap router edge cases) per the issue's "keep the existing fast mocked unit tests".

## #1581 — verified already implemented; pinned, no code (AC-runtime.5.1)

The cassette lookup has been input-keyed by construction since EPIC-023: `fingerprint = sha256(role + messages + decode_params)` (`src/llm/cassette.py`), and a replay miss is a **hard failure** (`test_AC23_6_2_replay_miss_is_hard_failure`, now cross-referenced as AC-runtime.5.1). The todo entry was a stale goal description.

## Governance

- 4 new roadmap ACs in `common/runtime/contract.py` (4.1 / 5.1 / 6.1 / 8.1), interface ↔ `__all__` parity for the 5 new checks + EPIC-007 references (doc-consistency check3/check5/check8 all green).
- `common/runtime/todo.md`: all enforcement items checked off with as-built notes; convergence snapshot now shows **no remaining gap column entries**.

## Verification

- Affected-suite regression: `tests/infra tests/runtime tests/api tests/llm tests/extraction tests/services` → **1335 passed**, 1 failed = `test_AC23_4_1_config_status_flips_when_user_configures`, which **fails identically on a clean main checkout** on this machine (local-env artifact, unrelated to this diff — CI is the arbiter).
- `tools/preflight.py`: ac-traceability / doc-consistency / api-reference / openapi-spec / tooling / frontend all ok; `backend-format` red is the known PATH-ruff 0.15 skew (pinned 0.14.11: check + format both clean).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tools/_lib/shell/smoke_test.sh [hunk 1]
```diff
@@ -193,6 +193,10 @@ check_endpoint "Journal Page" "$BASE_URL/journal" || FAILED=1
 check_endpoint "Statements Upload Page" "$BASE_URL/statements/upload" || FAILED=1
 check_endpoint "Reports Page" "$BASE_URL/reports" || FAILED=1
 check_endpoint "API Health" "$BASE_URL/api/health" "healthy" || FAILED=1
+# Dependency presence, manifest-driven (runtime invariant 6 / #1578): ?full=1
+# asserts every DEPENDENCY_MANIFEST.required_for(tier) dependency is present
+# for THIS deployment's tier; any absent declared dependency → 503 → smoke fails.
+check_endpoint "Dependency Presence (manifest, full)" "$BASE_URL/api/health?full=1" "healthy" || FAILED=1
```
[the remark this answers is not in the document]
Comment (wangzitian0) [reply tools/_lib/shell/smoke_test.sh]: Fixed in 322e1f26: all three health smoke checks now match `"status":"healthy"` (and the pinned contract string in `test_issue_489_deployment_contracts.py` is updated) — "unhealthy" can no longer false-positive.

File: apps/backend/src/runtime/extension/adapters.py [hunk 2]
```diff
@@ -139,3 +139,149 @@ async def probe(self) -> ProbeResult:
...
+
+        start = time.perf_counter()
+        try:
+            async with httpx.AsyncClient(
+                timeout=self._timeout_seconds, headers={"User-Agent": "finance-report-smoke/1.0"}
+            ) as client:
+                response = await client.get(self._PROBE_URL)
+            return ProbeResult(
+                self.name,
+                DependencyStatus.PRESENT,
```
[the remark this answers is not in the document]
Comment (wangzitian0) [reply apps/backend/src/runtime/extension/adapters.py]: Fixed in 322e1f26: `MarketDataCheck` now maps status ≥500 to ABSENT (429 stays PRESENT), aligned with the other HTTP probes; added `test_server_error_is_absent`.

File: apps/backend/src/main.py [hunk 3]
```diff
@@ -424,22 +432,34 @@ async def health_check(db: AsyncSession = Depends(get_db)) -> Response:
...
+
+        tier = None
+        if full:
+            tier = resolve_env_tier(
+                settings.environment, github_actions=os.getenv("GITHUB_ACTIONS", "").lower() == "true"
+            )
+            probed, unprobed = Bootloader._required_checks(tier)
+            for name in sorted(probed.keys() - {"database", "object_storage"}):
+                result = await getattr(Bootloader, probed[name])()
+                checks[name] = result.status == "ok"
```
[the remark this answers is not in the document]
Comment (wangzitian0) [reply apps/backend/src/main.py]: Fixed in 322e1f26: the ?full=1 probes run concurrently via `asyncio.gather` — latency is the slowest single probe (each has its own ~5s timeout and never raises).

File: apps/backend/tests/runtime/test_health_parity.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    response = await client.get("/health?full=1")
+
+    assert response.status_code == 200
+    body = response.json()
+    tier = EnvTier(body["tier"])
+    declared = DEPENDENCY_MANIFEST.required_for(tier)
+    # 'database' and legacy 's3' are the light-form keys; object_storage maps to s3.
+    reported = (set(body["checks"]) - {"s3"}) | ({"object_storage"} if "s3" in body["checks"] else set())
+    assert declared <= reported, f"declared {sorted(declared)} vs reported {sorted(reported)}"
```
[the remark this answers is not in the document]
Comment (wangzitian0) [reply apps/backend/tests/runtime/test_health_parity.py]: Fixed in 322e1f26: the parity assertion is strict set equality after the s3→object_storage mapping — an extra reported key now fails, exactly as invariant 6 demands.

File: apps/backend/tests/api/test_statement_pipeline_moto.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    stream = stream_ai_json(
+        messages=messages,
+        model=_TEXT_MODEL,
+        max_tokens=_TEXT_MAX_TOKENS,
+        temperature=0.0,
+        thinking={"type": "disabled"},
+    )
+    parsed = _loads_tolerant(await accumulate_stream(stream))
+    expected = json.loads(_EXPECTED.read_text(encoding="utf-8"))
+    assert parsed == expected
```
[the remark this answers is not in the document]
Comment (wangzitian0) [reply apps/backend/tests/api/test_statement_pipeline_moto.py]: Fixed in 322e1f26: the expected fixture stores balances/amounts as strings and the assertion compares Decimal-normalized values (str() before Decimal, per the repo money convention).

**Comment by wangzitian0** [comment]
Rebased onto main after #1601 landed the same #1520 territory (real StorageService over moto, AC8.25.1–.2, plus the reparse storage-key bug fix). This PR now **converges instead of duplicating**: the parallel moto test/fixture and AC-runtime.8.1 are dropped — runtime invariant 4 delegates to AC8.25.1–.2, mirroring how invariant 5 delegates to AC-llm.6.2. With no cassette-driving roadmap test left, the package tier is back to **CODE-ONLY** (reconcile: declared=CODE-ONLY detected=CODE-ONLY, share 0%). Scope now: #1580 probes + #1578 smoke parity + the delegation records for #1520/#1581.
